aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+template<CastPolicy policy, class Redispatch, Redispatch* F, class Ret, class ArgList> struct WrapFunction_ {};
+
+// Wrapper to infer return_type and parameter_types for WrapFunction_ (imitating core/boxing/kernel_function.h)
+template<CastPolicy policy,
+         class Registered, // The signature for which we're registering.  The dispatcher's calling code invokes our
+                           // registered functions with arguments matching Registered, so we register
+                           // WrapFunction_::call methods with a matching signature to properly field those arguments.
+                           // guts::function_traits below extracts return_type and parameter_types from Registered,
+                           // which WrapFunction_ templates above use to declare their call methods.
+         class Redispatch, // The signature for the function we're redispatching to.  In most cases this is the same
+                           // as Registered, but for some ops (for example, ops where we append a dtype) it's useful
+                           // to redispatch to a function with a different signature.
+         Redispatch* F>    // The actual function we're redispatching to.
+struct WrapFunction final {
+  using type = WrapFunction_<policy,
+                             Redispatch,
+                             F,
+                             typename guts::function_traits<Registered>::return_type,
+                             typename guts::function_traits<Registered>::parameter_types>;
+};
+
+// Aside from CastPolicy::fp16 and CastPolicy::fp32, the guts of each ::call below need divergent code.
+// Some need to flip a dtype, some need to append a dtype, some need to swap in a new casted vector in
+// place of a TensorList, etc, so I declare specializations.
+// I could deduplicate these by having a single template with more complex (one-size-fits-all) type selection and
+// TensorList handling, but with specializations, each instantiation can compile with only the logic it needs.
+
+// CastPolicy::fp16
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::fp16, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocasting(DispatchKey::AutocastTensorId);
+    return (*F)(cached_cast(at::kHalf, args)...);
+  }
+};
+
+// CastPolicy::fp16_with_tensorlist
+template<class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::fp16_with_tensorlist, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
+  static Ret call(Args... args) {
+    c10::impl::ExcludeDispatchKeyGuard no_autocasting(DispatchKey::AutocastTensorId);
+    // The TensorList that receives vec will use vec's raw data pointer.
+    // Create vec out of line to ensure its data's lifetime lasts the full duration of (*F)().
+    std::vector<Tensor> vec = get_casted_vector(at::kHalf, args...);
```

Comment:
Temporary will stay live for the entirety of the expression it's embedded in, so technically this doesn't need to happen externally

Reply:
If the receiving call were accepting an `std::vector&` reference to the full thing, I'd agree, but the TensorList that accepts vec takes a blind reference to its data pointer.  I'm not sure if that's sufficient to extend vec's lifetime.  Can you point me to some cppreference that supports your claim?  I'm not trying to be argumentative, genuinely curious to hear your justification, because this stuff is tricky and the standard-ese is often unclear to me without an example.

In the meantime I decided to be safe rather than sorry by creating vec out of line.

Reply:
You're thinking of const-reference extension. If I have:

```
const A& x = foo(); // A foo();
...use of x ...
```

The fact that the temporary is assigned to a const reference extends its lifetime for further uses of x. If you replaced `const A&` with `ArrayRef`, for example, const reference extension would no longer apply. This is a major hazard of view objects. https://abseil.io/tips/107

Here's the reference on lifetime of temporaries:  https://en.cppreference.com/w/cpp/language/lifetime  The relevant verbiage is: "All temporary objects are destroyed as the last step in evaluating the full-expression that (lexically) contains the point where they were created"

Reply:
Sorry, I meant `const std::vector&` above, edited.  I stumbled on that same abseil post when I was writing the TensorList treatment; the fact that ArrayRef takes a view is exactly the source of my concern.

The lifetime rule that you say saves me appears to be something separate:  if I create vec as a temporary inline, its lifetime won't be extended because ArrayRef happens to take a view.  Rather, vec is destined to last until the semicolon anyway, because that's just the way things are.  Am I interpreting correctly?

I'll try creating vec inline and see if i get segfaults ¯\_(ツ)_/¯.  If it works, I'll get rid of the divergent `*_with_tensorlist` CastPolicies (unify them with the simple fp16 and fp32 case as I said [earlier](https://github.com/pytorch/pytorch/pull/32140#discussion_r375605321)).

Reply:
> Rather, vec is destined to last until the semicolon anyway, because that's just the way things are. Am I interpreting correctly?

Yes.

> I'll try creating vec inline and see if i get segfaults ¯_(ツ)_/¯.

A in fact terrible way to do C++ development :3

Reply:
https://i.imgflip.com/2qerll.jpg

yes, this is tricky, easy to know if we get it wrong but hard to prove we got it right.  If i'm requesting undefined behavior from the compiler, i might get lucky and it will work.  But if vec should last til the semicolon, we're ok.

Reply:
I rewrote TensorList handling to create the vec temporary directly in the arglist.   `cached_cast` has an overload that accepts a TensorList and creates a vector temporary, so`(*F)(cached_cast(at::kFloat, args)...);` can handle `args` that may contain TensorLists.   This means the forwarding doesn't need divergent code for args with a TensorList, so I got rid of `fp32_with_tensorlist` and `fp16_with_tensorlist` as separate policies (functions with TensorLists are now handled by the ordinary `fp32` and `fp16` policies).